attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,126 @@
+    pub provider: String,
+}
+
+impl TsmReportPath {
+    pub fn open() -> Result<Self, TsmReportError> {
+        let p = tempdir_in(TSM_REPORT_PATH).map_err(|e| TsmReportError::Open(e))?;
+
+        let path = p.into_path();
+
+        let provider = match std::fs::read_to_string(path.as_path().join("provider")) {
+            Ok(tsm) => tsm.trim_matches('\n').to_string(),
+            Err(e) => {
+                let _ = std::fs::remove_dir(path.as_path())
+                    .map_err(|r| log::error!("Failed to remove TSM Report directory: {}", r));
+                return Err(TsmReportError::Access("provider", e));
+            }
+        };
+
+        Ok(Self { path, provider })
+    }
+    pub fn close(&self) {
+        let _ = std::fs::remove_dir(self.path.as_path())
+            .map_err(|e| log::error!("Failed to remove TSM Report directory: {}", e));
+    }
+    pub fn attestation_report(
+        &self,
+        provider: TsmReportProvider,
+    ) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let report_data = match provider {
+            TsmReportProvider::Tdx(inblob) => inblob,
+            TsmReportProvider::Sev(privlevel, inblob) => {
+                // TODO: untested
+                std::fs::write(report_path.join("privlevel"), vec![privlevel])
+                    .map_err(|e| TsmReportError::Access("privlevel", e))?;
+                inblob
+            }
+        };
+
+        if report_data.len() == 0 {
+            return Err(TsmReportError::InblobLen);
+        }
+
+        std::fs::write(report_path.join("inblob"), report_data)
+            .map_err(|e| TsmReportError::Access("inblob", e))?;
+
+        let q = std::fs::read(report_path.join("outblob"))
+            .map_err(|e| TsmReportError::Access("outblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(q)
+    }
+    pub fn supplemental_data(&self) -> Result<Vec<u8>, TsmReportError> {
+        let report_path = self.path.as_path();
+
+        let aux = std::fs::read(report_path.join("auxblob"))
+            .map_err(|e| TsmReportError::Access("auxblob", e))?;
+
+        check_inblob_write_race(report_path)?;
+
+        Ok(aux)
+    }
+}
+
+fn check_inblob_write_race(report_path: &std::path::Path) -> Result<(), TsmReportError> {
+    let g = std::fs::read_to_string(report_path.join("generation"))
+        .map_err(|e| TsmReportError::Access("generation", e))?;
+
+    let generation = g.trim_matches('\n').to_string().parse::<u32>().unwrap();
+
+    if generation > 1 {
+        return Err(TsmReportError::InblobConflict(generation));
+    }
+
+    Ok(())
+}
+
+pub fn provider_is(wanted: &str) -> bool {
+    if !Path::new(TSM_REPORT_PATH).exists() {
+        return false;
+    }
+
+    match TsmReportPath::open() {
+        Ok(report_path) => {
+            let provider = report_path.provider.clone();
+            report_path.close();
+            provider.cmp(&wanted.to_string()).is_eq()
```

Comment:
Could we directly use
```suggestion
            provider == wanted
```